mllib/src/main/scala/org/apache/spark/ml/regression/GeneralizedLinearRegression.scala

```diff
@@ -0,0 +1,472 @@
+
+  override def initialize(y: Double, weight: Double): Double = y
+
+  def variance(mu: Double): Double = 1.0
+}
+
+private[ml] object Gaussian {
+
+  def apply(link: String): Gaussian = {
+    link match {
+      case "identity" => new Gaussian(Identity)
+      case "log" => new Gaussian(Log)
+      case "inverse" => new Gaussian(Inverse)
+    }
+  }
+}
+
+/**
+ * Binomial exponential family distribution.
+ * The default link for the Binomial family is the logit link.
+ * @param link a link function instance
+ */
+private[ml] class Binomial(link: Link = Logit) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = {
+    (weight * y + 0.5) / (weight + 1.0)
+  }
+
+  override def variance(mu: Double): Double = mu * (1 - mu)
+}
+
+private[ml] object Binomial {
+
+  def apply(link: String): Binomial = {
+    link match {
+      case "logit" => new Binomial(Logit)
+      case "probit" => new Binomial(Probit)
+      case "cloglog" => new Binomial(CLogLog)
+    }
+  }
+}
+
+/**
+ * Poisson exponential family distribution.
+ * The default link for the Poisson family is the log link.
+ * @param link a link function instance
+ */
+private[ml] class Poisson(link: Link = Log) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = y + 0.1
+
+  override def variance(mu: Double): Double = mu
+}
+
+private[ml] object Poisson {
+
+  def apply(link: String): Poisson = {
+    link match {
+      case "log" => new Poisson(Log)
+      case "sqrt" => new Poisson(Sqrt)
+      case "identity" => new Poisson(Identity)
+    }
+  }
+}
+
+/**
+ * Gamma exponential family distribution.
+ * The default link for the Gamma family is the inverse link.
+ * @param link a link function instance
+ */
+private[ml] class Gamma(link: Link = Log) extends Family(link) {
+
+  override def initialize(y: Double, weight: Double): Double = y
+
+  override def variance(mu: Double): Double = math.pow(mu, 2.0)
+}
+
+private[ml] object Gamma {
+
+  def apply(link: String): Gamma = {
+    link match {
+      case "inverse" => new Gamma(Inverse)
+      case "identity" => new Gamma(Identity)
+      case "log" => new Gamma(Log)
+    }
+  }
+}
+
+/**
+ * A description of the link function to be used in the model.
+ */
+private[ml] trait Link extends Serializable {
+
+  /** The link function. */
+  def link(mu: Double): Double
+
+  /** The derivative function. */
+  def deriv(mu: Double): Double
+
+  /** The inverse link function. */
+  def unlink(eta: Double): Double
+}
+
+private[ml] object Identity extends Link {
+
+  override def link(mu: Double): Double = mu
+
+  override def deriv(mu: Double): Double = 1.0
+
+  override def unlink(eta: Double): Double = eta
+}
+
+private[ml] object Logit extends Link {
+
+  override def link(mu: Double): Double = math.log(mu / (1.0 - mu))
+
+  override def deriv(mu: Double): Double = 1.0 / (mu * (1.0 - mu))
+
+  override def unlink(eta: Double): Double = 1.0 / (1.0 + math.exp(-1.0 * eta))
+}
```

Comment:
I think the restriction on endogenous variable should go into the `Family` class since it is truly the distribution on Y that restricts the values. This is how R does it.